pcdbapi/main.py

```diff
@@ -0,0 +1,160 @@
+
+from middleware import encode_response, verify_signature
+
+# All config in one place
+CONFIG = {
+    "mongodb_url": os.getenv("MONGODB_URL"),
+    "api_secret": os.getenv("API_SECRET"),  # shared secret
+    "max_timestamp_diff": 300,  # 5 minutes
+}
+
+
+class OidcClient(BaseModel):
+    model_config = ConfigDict(extra="forbid")
+
+    # These are the only fields that can be set through the API
+    name: constr(min_length=1, max_length=200) | None = None
+    redirect_uris: List[str] | None = None
+    post_logout_redirect_uris: List[str] | None = None
+    id_token_signed_response_alg: Literal["RS256", "ES256", "HS256"] | None = None
+    userinfo_signed_response_alg: Literal["RS256", "ES256", "HS256"] | None = None
+    active: bool | None = None
+
+
+def validate_objectid(_id: str) -> ObjectId:
+    try:
+        oid = ObjectId(_id)
+    except:
+        raise HTTPException(status_code=422, detail="Invalid ObjectId") from None
+    return oid
+
+
+@asynccontextmanager
+async def lifespan(app: FastAPI):
+    # Startup
+    app.mongodb_client = AsyncIOMotorClient(CONFIG["mongodb_url"])
+    app.db = app.mongodb_client.get_default_database()
+    app.collection = app.db.oidc_clients
+    # Wait for database to be ready
+    for _ in range(10):
+        try:
+            ping_response = await app.db.command("ping")
+            break
+        except:  # pragma: no cover
+            await asyncio.sleep(1)
+    if int(ping_response["ok"]) != 1:  # pragma: no cover
+        raise Exception("Problem connecting to database.")
+    yield
+    # Shutdown
+    app.mongodb_client.close()
+
+
+app = FastAPI(lifespan=lifespan)
+
+# Middleware to verify signature on /api/*
+app.middleware("http")(lambda request, call_next: verify_signature(request, call_next, CONFIG))
+
+
+@app.get("/healthz")
+async def healthz():
+    ping_response = await app.db.command("ping")
+    if int(ping_response["ok"]) != 1:  # pragma: no cover
+        raise HTTPException(status_code=500, detail="Database connection failed")
+    return "ok"
+
+
+@app.get("/api/oidc_clients")
+@encode_response
+async def list_oidc_clients(request: Request):
+    cursor = app.collection.find({"email": request.state.email})
+    elts = await cursor.to_list(None)
+    return elts
+
+
+@app.post("/api/oidc_clients")
+@encode_response
+async def create_oidc_client(data: OidcClient, request: Request):
+    d = data.model_dump(exclude_unset=True)
+    d.update(
+        {
+            "email": request.state.email,
+            "createdAt": datetime.now(),
+            "updatedAt": datetime.now(),
+            "updatedBy": "espace-partenaires",
+
+            # TODO: fix these fields?
+            "title": "Nouvelle application",
+            "site": ["https://site.com"],
+
+            # Generate IDs in correct format
+            "key": secrets.token_hex(32),  # 64 hex chars
+            "client_secret": secrets.token_hex(32),  # 64 hex chars
```

Comment:
This field [must be encrypted](https://github.com/france-connect/sources/blob/main/back/libs/service-provider-adapter-mongo/src/service-provider-adapter-mongo.service.ts#L163) using the [`CLIENT_SECRET_CIPHER_PASS` env var](https://github.com/france-connect/sources/blob/main/back/instances/core-fca-low/src/config/service-provider-adapter-mongo.ts#L7) 

The decryption algorithm [can be found here](https://github.com/france-connect/sources/blob/main/back/libs/cryptography/src/cryptography.service.ts#L112)

Reply:
https://github.com/numerique-gouv/espace-partenaire/commit/842c9b2b63a83a0536882defd32142fddd7a7123